src/Stack/PackageIndex.hs

```diff
@@ -247,12 +248,19 @@ updateIndexGit menv indexName' index gitUrl = do
                   $(mkRelDir "git-update")
                 acfDir = suDir </> repoName
             repoExists <- doesDirExist acfDir
-            unless repoExists
-                   (readProcessNull (Just suDir) menv "git" cloneArgs)
+            let doClone = readProcessNull (Just suDir) menv "git" cloneArgs
+            unless repoExists doClone
             isShallow <- doesFileExist $ acfDir </> $(mkRelDir ".git") </> $(mkRelFile "shallow")
             when isShallow $ do
               $logWarn "Shallow package index repo detected, transitioning to a full clone..."
-              readProcessNull (Just acfDir) menv "git" ["fetch", "--unshallow"]
+              let handleUnshallowError =
+                    C.handle $ \case
+                      ReadProcessException{} -> do
+                        $logInfo $ "Failed to convert to full clone, deleting and re-cloning."
+                        ignoringAbsence (removeDirRecur acfDir)
+                        doClone
+                      err -> throwM err
```

Comment:
`ReadProcessException` is one data constructor of for in type `ReadProcessException`, so this catch all prevents a exhaustivity warning (which risks a match error at runtime). IIUC you assumed `ReadProcessException` did not have alternatives, is that because it violates some naming pattern?
```
Warning:
    Pattern match(es) are non-exhaustive
    In a case alternative:
        Patterns not matched:
            NoPathFound
            ExecutableNotFound _ _
            ExecutableNotFoundAt _
```